internal/merkle/proof.go

```diff
@@ -0,0 +1,100 @@
+// (c) Cartesi and individual authors (see AUTHORS)
+// SPDX-License-Identifier: Apache-2.0 (see LICENSE)
+
+package merkle
+
+import (
+	"errors"
+
+	"github.com/cartesi/rollups-node/internal/node/model"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+// CreateProofs creates proofs for all the leaves of the binary Merkle tree
+// with the given height. It returns the root hash and the siblings matrix,
+// in bottom-up order.
+//
+// If the number of leaves exceeds the capacity for the given height,
+// an error is returned.
+func CreateProofs(leaves []model.Hash, height uint) (model.Hash, [][]model.Hash, error) {
+	pristineNode := model.Hash{}
+
+	if len(leaves) == 0 {
+		leaves = append(leaves, pristineNode)
+	}
+
+	currentLevel := leaves
+	nextLevel := make([]model.Hash, (len(leaves)+1)/2)
+	siblings := make([][]model.Hash, len(leaves))
+
+	// for each level in the tree, starting from the leaves
+	for levelIdx := 0; levelIdx < int(height); levelIdx++ {
+		calculateSiblings(levelIdx, int(height), currentLevel, siblings, pristineNode)
```

Comment:
It was a workaround as Golang requires both operands to be the same type. For brevity, I preferred `height` cast to int instead of `levelIdx` to uint, but converting `calculateSiblings` parameters to unsigned ints is better 👍🏽

Reply:
I didn't know Golang didn't coerce 0 into `uint`. You could do this then...

```go
	for levelIdx := uint(0); levelIdx < height; levelIdx++ {
```